azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -59,95 +60,164 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
+            try:
+                uri = findtext(self._xml_doc, "SharedConfig")
+                if uri is None:
+                    logger.warn("SharedConfig url doesn't exist in goal state")
+                else:
+                    xml_text = self._wire_client.fetch_config(uri, self._wire_client.get_header())
+                    self._shared_conf = SharedConfig(xml_text)
+                self.shared_conf_retrieved = True
+            except Exception as e:
+                self._log_document_retrieval_failure("shared config", e)
+                raise
+        return self._shared_conf
+
+    @property
+    def remote_access(self):
+        if not self.remote_access_retrieved:
+            try:
+                container = find(self._xml_doc, "Container")
+                uri = findtext(container, "RemoteAccessInfo")
+                if uri is None:
+                    self._remote_access = None
+                else:
+                    xml_text = self._wire_client.fetch_config(uri, self._wire_client.get_header_for_cert())
+                    self._remote_access = RemoteAccess(xml_text)
+                self.remote_access_retrieved = True
+            except Exception as e:
+                self._log_document_retrieval_failure("remove access", e)
+                raise
+        return self._remote_access
+
+    @property
+    def ext_conf(self):
+        if not self._ext_conf_retrieved:
+            try:
+                uri = findtext(self._xml_doc, "ExtensionsConfig")
+                self._ext_conf = self._ext_config_retriever.get_ext_config(self.incarnation, uri)
+                self._ext_conf_retrieved = True
+            except Exception as e:
+                self._log_document_retrieval_failure("extensions config", e)
+                raise
+        return self._ext_conf
+
+    @property
+    def artifacts_profile_blob_url(self):
+        if not self._ext_conf_properties_retrieved:
+            self._retrieve_fabric_ext_conf_properties()
+        return self._artifacts_profile_blob_url
+
+    @property
+    def status_upload_blob_url(self):
+        if not self._ext_conf_properties_retrieved:
+            self._retrieve_fabric_ext_conf_properties()
+        return self._status_upload_blob_url
+
+    @property
+    def status_upload_blob_type(self):
+        if not self._ext_conf_properties_retrieved:
+            self._retrieve_fabric_ext_conf_properties()
+        return self._status_upload_blob_type
+
+    def _log_document_retrieval_failure(self, component_name, e):
+        estr = ustr(e)
+        if self._HttpFailedIndicator in estr:
+            logger.warn("Fetching the {0} failed: {1}", component_name, ustr(e))
+        else:
+            logger.warn("Fetching the {0} failed: {1}, {2}", component_name, ustr(e), traceback.format_exc())
+
+    def _retrieve_fabric_ext_conf_properties(self):
+        try:
+            # The artifacts profile blob url and status blob url are only in the Fabric goal state
```

Comment:
Just confirming here - in case the artifacts profile blob changes, we would need to process fabric goal state first since the fast track channel would be invalidated. Luckily, if there is an artifact profile blob change, that is a fabric GS anyway (as opposed to fast track) since it contains updates other than related to extensions. Is this correct?

Reply:
If the VmArtifactsProfileBlob changes, then we need to process the FastTrack GS (unless the Fabric incarnation also changed). This code is avoiding a chicken and egg scenario.
- We don't know whether to process the Fabric GS or the FastTrack GS until we have both
- We can't retrieve the FastTrack GS until we have the Fabric GS, because it has the urls
- So, we retrieve the urls from the Fabric GS in order to proceed.